### PR TITLE
OneWeek2018

### PR DESCRIPTION
**Pull request by rabee333**
Including the following items:
#18 App AuthN with certificate
#14 Interactive Login option
#13 Re-design AzureStorageHelper
#5 Support device login
And few more fixes.

### CODE REVIEW

File: data/src/test/java/com/microsoft/azure/kusto/data/AadAuthenticationHelperTest.java [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+        X509Certificate x509Certificate = readPem(certFilePath, "basic").getCertificate();
+        PrivateKey privateKey = readPem(privateKeyPath, "basic").getKey();
+
+        ConnectionStringBuilder csb = ConnectionStringBuilder
+                .createWithAadApplicationCredentials("resource.uri", "client-id", "applicationKey", "authority-id");
+
+        AadAuthenticationHelper aadAuthenticationHelper = new AadAuthenticationHelper(csb);
+
+        Assertions.assertThrows(NullPointerException.class,
+                () -> aadAuthenticationHelper.acquireWithClientCertificate(x509Certificate, null));
```
Comment (LizaShak) [inline data/src/test/java/com/microsoft/azure/kusto/data/AadAuthenticationHelperTest.java]: add to connectionStringBuilder to verify null params

File: ingest/src/main/java/com/microsoft/azure/kusto/ingest/source/SourceInfo.java [hunk 2]
```diff
@@ -1,17 +1,5 @@
...
-    public UUID getSourceId() {
-        return sourceId;
-    }
-
-    public void setSourceId(UUID sourceId) {
-        this.sourceId = sourceId;
-    }
-
+public interface SourceInfo {
+    void validate();
```
Comment (LizaShak) [inline ingest/src/main/java/com/microsoft/azure/kusto/ingest/source/SourceInfo.java]: Guid Id

Comment (tamirkamara) [reply ingest/src/main/java/com/microsoft/azure/kusto/ingest/source/SourceInfo.java]: @rabee333 I think the meaning here is that the interface should have the get/set for the sourceId and the abstract class will have the implementation (very basic) as well as the actual field

File: data/src/test/java/com/microsoft/azure/kusto/data/AadAuthenticationHelperTest.java (hunk 1 above)
Comment (LizaShak) [reply data/src/test/java/com/microsoft/azure/kusto/data/AadAuthenticationHelperTest.java]: I added this validation for other functions as well.

File: data/src/main/java/com/microsoft/azure/kusto/data/AadAuthenticationHelper.java [hunk 3]
```diff
@@ -22,36 +25,64 @@
...
     }
 
-    String acquireAccessToken() throws DataServiceException, DataClientException {
-        if (clientCredential != null) {
-            return acquireAadApplicationAccessToken().getAccessToken();
-        } else {
-            return acquireAadUserAccessToken().getAccessToken();
+    String acquireAccessToken() throws DataServiceException  {
+        try {
+            switch (authenticationType) {
```
Comment (toshetah) [inline data/src/main/java/com/microsoft/azure/kusto/data/AadAuthenticationHelper.java]: Aren't we missing here AAD_APPLICATION_CERTIFICATE?

File: data/src/main/java/com/microsoft/azure/kusto/data/AadAuthenticationHelper.java [hunk 4]
```diff
@@ -95,7 +126,72 @@ private AuthenticationResult acquireAadApplicationAccessToken() throws DataServi
         return result;
     }
 
-    private Boolean isNullOrEmpty(String str) {
-        return (str == null || str.trim().isEmpty());
+    public AuthenticationResult acquireAccessTokenUsingDeviceCodeFlow() throws Exception {
```
Comment (toshetah) [inline data/src/main/java/com/microsoft/azure/kusto/data/AadAuthenticationHelper.java]: Can we make this method private?

File: data/src/main/java/com/microsoft/azure/kusto/data/AadAuthenticationHelper.java [hunk 5]
```diff
@@ -95,7 +126,72 @@ private AuthenticationResult acquireAadApplicationAccessToken() throws DataServi
...
+        }
+        return result;
+    }
+
+    private AuthenticationResult waitAndAcquireTokenByDeviceCode(DeviceCode deviceCode, AuthenticationContext context)
+            throws  InterruptedException{
+        int timeout = 15 * 1000;
+        AuthenticationResult result = null;
+        while (timeout > 0){
+            try{
```
Comment (toshetah) [inline data/src/main/java/com/microsoft/azure/kusto/data/AadAuthenticationHelper.java]: IIUC, ExecutionException can be raised from the Future.get iff the context.acquireTokenByDeviceCode threw an exception. It that's the case, is there any value of retrying it?
Moreover, the get() should be blocking, so I am not sure there is any value of that busy wait.
Am I missing something?

Comment (LizaShak) [reply data/src/main/java/com/microsoft/azure/kusto/data/AadAuthenticationHelper.java]: This is the type of Exception thrown when it is still pending for user authentication. So the wait is not  for future result, but for user to finish the authentication.

**Pull request merged**